_posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd

```diff
@@ -0,0 +1,705 @@
+
+## Brain image segmentation
+
+With U-Net, domain applicability is as broad as the architecture is flexible. Here, we want to detect abnormalities in brain scans. The dataset, used in @BUDA2019218, contains MR images together with manually created [FLAIR](https://en.wikipedia.org/wiki/Fluid-attenuated_inversion_recovery) abnormality segmentation masks. It is available on [Kaggle](https://www.kaggle.com/mateuszbuda/lgg-mri-segmentation).
+
+Nicely, the paper is accompanied by a [GitHub repository](https://github.com/mateuszbuda/brain-segmentation-pytorch). Below, we closely follow (though not exactly replicate) the authors' preprocessing and data augmentation code.
+
+As is often the case in medical imaging, there is notable class imbalance in the data. For every patient, sections have been taken at multiple positions. (Number of sections per patient varies.) Most sections do not exhibit any lesions; the corresponding masks are colored black everywhere.
+
+Here are three examples where the masks *do* indicate abnormalities:
+
+![](images/scans.png)
+
+Let's see if we can build a U-Net that generates such masks for us.
+
+## Data
+
+Before you start typing, here is a [Colaboratory notebook](https://colab.research.google.com/drive/1gcPOlF4jff42yiquRbbCxnRANEqPpGMD?usp=sharing) to conveniently follow along.
+
+We use `pins` to obtain the data. Please see [this introduction](https://pins.rstudio.com/articles/boards-kaggle.html) if you haven't been using that package before.
+
+```{r}
+library(torch)
+library(torchvision)
+library(tidyverse)
+library(magick)
+library(zeallot)
+library(cowplot)
+library(zip)
+library(pins)
+
+torch_manual_seed(777)
+set.seed(777)
+
+# use your own kaggle.json here
+pins::board_register_kaggle(token = "~/kaggle.json")
+
+files <- pins::pin_get("mateuszbuda/lgg-mri-segmentation", board = "kaggle",  extract = FALSE)
+
+
+```
+
+The dataset is not that big -- it includes scans from 110 different patients --, so we'll have to do with just a training and a validation set. (Don't do this in real life, as you'll inevitably end up fine-tuning on the latter.)
+
+```{r}
+train_dir <- "data/mri_train"
+valid_dir <- "data/mri_valid"
+
+if(dir.exists(train_dir)) unlink(train_dir, recursive = TRUE, force = TRUE)
+if(dir.exists(valid_dir)) unlink(valid_dir, recursive = TRUE, force = TRUE)
+
+zip::unzip(files, exdir = "data")
+
+file.rename("data/kaggle_3m", train_dir)
+
+# this is a duplicate, again containing kaggle_3m (evidently a packaging error on Kaggle)
+# we just remove it
+unlink("data/lgg-mri-segmentation", recursive = TRUE)
+
+dir.create(valid_dir)
+
+```
+
+Of those 110 patients, we keep 30 for validation. Some more file manipulations, and we're set up with a nice hierarchical structure, with each of `train_dir` and `valid_dir` holding their per-patient sub-directories, respectively.
```

Comment:
```suggestion
Of those 110 patients, we keep 30 for validation. Some more file manipulations, and we're set up with a nice hierarchical structure, with each of `train_dir` and `valid_dir` holding its per-patient sub-directories, respectively.
```

Reply:
Alternatively, you could remove the each, so it would read something like:
> with `train_dir` and `valid_dir` holding their per-patient sub-directories, respectively.

Reply:
oh, that one's great, will do